modules/pool/lib/pool-onchain-data.ts

```diff
@@ -0,0 +1,211 @@
+import LiquidityBootstrappingPoolAbi from '../abi/LiquidityBootstrappingPool.json';
+import ComposableStablePoolAbi from '../abi/ComposableStablePool.json';
+import GyroEV2Abi from '../abi/GyroEV2.json';
+import VaultAbi from '../abi/Vault.json';
+import aTokenRateProvider from '../abi/StaticATokenRateProvider.json';
+import WeightedPoolAbi from '../abi/WeightedPool.json';
+import StablePoolAbi from '../abi/StablePool.json';
+import MetaStablePoolAbi from '../abi/MetaStablePool.json';
+import StablePhantomPoolAbi from '../abi/StablePhantomPool.json';
+import { JsonFragment } from '@ethersproject/abi';
+
+const abi: JsonFragment[] = Object.values(
+    // Remove duplicate entries using their names
+    Object.fromEntries(
+        [
+            ...ElementPoolAbi,
+            ...LinearPoolAbi,
+            ...LiquidityBootstrappingPoolAbi,
+            ...ComposableStablePoolAbi,
+            ...GyroEV2Abi,
+            ...VaultAbi,
+            ...aTokenRateProvider,
+            ...WeightedPoolAbi,
+            ...StablePoolAbi,
+            ...StablePhantomPoolAbi,
+            ...MetaStablePoolAbi,
+            ...ComposableStablePoolAbi,
+            //...WeightedPoolV2Abi,
+        ].map((row) => [row.name, row]),
+    ),
+);
+
+const getSwapFeeFn = (type: string) => {
+    if (type === 'ELEMENT') {
+        return 'percentFee';
+    } else if (type === 'FX') {
+        return 'protocolPercentFee';
+    } else {
+        return 'getSwapFeePercentage';
+    }
+};
+
+interface PoolInput {
+    id: string;
+    address: string;
+    type: PrismaPoolType;
+    tokens: {
+        address: string,
+        token: {
+            decimals: number,
+        }
+    }[];
+    version?: number;
+}
+
+interface OnchainData {
+    poolTokens: [string[], BigNumber[]];
+    totalSupply: BigNumber;
+    virtualSupply?: BigNumber;
+    actualSupply?: BigNumber;
+    swapFee: BigNumber;
+    swapEnabled?: boolean;
+    protocolYieldFeePercentageCache?: BigNumber;
+    rate?: BigNumber;
+    weights?: BigNumber[];
+    targets?: [BigNumber, BigNumber];
+    wrappedTokenRate?: BigNumber;
+    amp?: [BigNumber, boolean, BigNumber];
+    tokenRates?: [BigNumber, BigNumber];
+    tokenRate?: BigNumber[];
+    metaPriceRateCache?: [BigNumber, BigNumber, BigNumber][];
+}
+
+const defaultCalls = (
+    { id, address, type }: PoolInput,
+    vaultAddress: string,
+    multicaller: Multicaller3
+) => {
+    multicaller.call(`${id}.poolTokens`, vaultAddress, 'getPoolTokens', [id]);
+    multicaller.call(`${id}.totalSupply`, address, 'totalSupply');
+    multicaller.call(`${id}.virtualSupply`, address, 'getVirtualSupply');
+    multicaller.call(`${id}.actualSupply`, address, 'getActualSupply');
+    multicaller.call(`${id}.swapFee`, address, getSwapFeeFn(type));
+    multicaller.call(`${id}.rate`, address, 'getRate');
+    multicaller.call(`${id}.protocolYieldFeePercentageCache`, address, 'getProtocolFeePercentageCache', [2]);
```

Comment:
Same here, but I dont see a way to have this without relying on the network context :)